src/core/hooks/useState.ts

```diff
@@ -0,0 +1,46 @@
+import { currentComponent } from "@/core";
+import { render } from "@/index";
+
+interface SetState<T> {
+  (value: T | ((previousState: T) => T)): void;
+}
+
+const useState = (() => {
+  const componentsState: Record<string, unknown[]> = {};
+
+  return function <T>(initialValue: T): [T, SetState<T>] {
+    if (!currentComponent) {
+      throw new Error("useState는 컴포넌트 안에서만 호출될 수 있습니다.");
+    }
+
+    const { id, stateIndex } = currentComponent;
+
+    if (!componentsState[id]) {
+      componentsState[id] = [];
+    }
+
+    if (componentsState[id][stateIndex] === undefined) {
+      componentsState[id][stateIndex] = initialValue;
+    }
+
+    const setState: SetState<T> = (newValue) => {
+      const currentState = componentsState[id][stateIndex] as T;
+
+      const updatedState =
+        typeof newValue === "function" ? (newValue as (previousValue: T) => T)(currentState) : newValue;
+
+      if (currentState !== updatedState) {
+        componentsState[id][stateIndex] = updatedState;
+        render();
+      }
+    };
```

Comment:
전에도 말씀드려서 이미 인지하고 계실 것 같지만 타입 스크립트에서 as를 사용하는 타입 단언은 실행 시점의 에러는 해결이 가능하지만 계속해서 타입 단언을 사용할 수 밖에 없게 되는 문제도 있고, 사실상 타입스크립트의 기능을 사용했다기보다 개발자 본인이 해당 타입에 대해 확신을 하고 지정을 하는 것이기 때문에 타입스크립트 활용면에 있어서는 아쉽다고 생각이 들어요.

그래서 혹시 가능하다면 타입 가드를 추가해보시는 것이 어떠신가 해서 아래와 같이 제안드립니다. 

```suggestion
if (!(componentsState[id][stateIndex] instanceof Object)) {
  componentsState[id][stateIndex] = initialValue;
}

const setState: SetState<T> = (newValue) => {
  const currentState = componentsState[id][stateIndex] as T;

  const updatedState =
    typeof newValue === "function" ? (newValue as (previousValue: T) => T)(currentState) : newValue;

  if (currentState !== updatedState) {
    if (updatedState instanceof Object) {
      componentsState[id][stateIndex] = { ...currentState, ...updatedState };
    } else {
      componentsState[id][stateIndex] = updatedState;
    }
    render();
  }
};
```

타입 가드가 어쩔수없이 타입 단언보다는 길게 써야한다는 점에서 가독성이 떨어진다는 점이 아쉽지만 보다 확실한 타입 검사를 위해서 제안 드려요 ㅠ
(위에 제안 드린 코드는 맞을지는 모르겠어요. 그냥 예시로 제안드리는 것이니 참고 부탁드립니다.)

Reply:
말씀해주신 부분에 대해서 타입 가드 함수를 접목하여(`is`를 활용하여) 시도해보았으나 완전히 타입 단언을 제거하기가 까다로웠습니다. 제안해주신 방법은 객체일때와 아닐때의 분기처리를 해주신 것 같은데 이 방법만으로는 타입 단언을 제거하기 어려울 것 같습니다. 
컴포넌트가 앞으로 사용하게될 상태의 타입을 추론하는 것이 불가능하고 상태를 `string | number | boolean | obect | array` 이런 식으로 설정한다면 결국 unknown과 다를게 없다는 생각이 들었습니다.
관련해서 혹시 다른 의견이 있으시다면 말씀해주세요 : )

Reply:
먼저 타입 단언이 발생하는 근본적인 이유는 `componentState` 객체의 값이 `unknown`으로 설정되어 있기 때문입니다. `componentState`는 여러 컴포넌트의 상태를 저장하게 되고 예를 들면 다음과 같습니다.

```ts
{ "ComponentA": ["id", 1, ["todos"]] };
{ "ComponentB": ["some state"] };
{ "ComponentC": [12] };
```

컴포넌트는 내부적으로 다양한 데이터 타입을 상태로 관리할 수 있습니다. useState를 사용하여 초기 상태를 설정할 때, 내부적으로 상태의 타입을 결정합니다. 그렇기 때문에, `componentState` 내에서 `unknown` 대신 구체적인 타입을 사용하면, setState에서의 타입 단언을 피할 수 있지만, 모든 타입에 대해 추상화된 형태로 useState를 사용할 수는 없게 됩니다.

그래서 저 역시 타입 단언의 위험성을 인식하고 있고 여러 시도를 해보았지만 타입 단언을 제거할 수 없었습니다.


@InSeong-So 님
> 분기를 치기 어렵다면 타입별로 함수든, 컴포넌트든 생성하는 것이 안정성 측면에서 더 좋은 것을 이해해주면 좋겠습니다.

리뷰 주신 이 부분에 대해서는 useState와 유사하게 작동하지만 특정 타입에 특화된 `useStringState`, `useNumberState` 등을 구현하여 사용하라는 것으로 이해했는데 맞게 이해한걸까요?

이렇게 하면 `componentState`의 값에 대한 타입을 명시적으로 지정할 수 있어 setState 함수 내부에서 타입 단언이 필요 없어지긴 하는데요.
대신 각 타입별로 별도의 useState 함수를 생성하여 사용하는 것이 반드시 필요한지에 대해서는 고민이 됩니다. 혹시 이에 대한 의견을 들을 수 있을까요?

혹은 저 코드 상황에서 타입 단언을 제거할 수 있는 예시를 좀 들어주시면 너무 감사하겠습니다.. 계속 말씀드렸지만 타입 단언에 대한 위험성도 인지하고 있고 최대한 지양하려 하지만 잘 안되네요.. 😭